usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -2697,95 +2772,73 @@ static CK_RV make_maced_spki(STDLL_TokData_t *tokdata, SESSION * sess,
-                TRACE_ERROR("%s\n", ock_err(ERR_ATTRIBUTE_VALUE_INVALID));
-                rc = CKR_ATTRIBUTE_VALUE_INVALID;
+        /*
+         * m_UnwrapKey with CKM_IBM_TRANSPORTKEY allows boolean attributes only
+         * to be added to MACed-SPKIs
+         */
+        node = pub_key_obj->template->attribute_list;
+        while (node != NULL) {
+            rc = check_add_spki_attr(tokdata, node->data, keytype, curve_type,
+                                     &p_attrs, &attrs_len);
+            if (rc != CKR_OK)
                 goto make_maced_spki_end;
-            }
 
-            /*
-             * EP11 does not allow to restrict public RSA/DSA/EC keys with
-             * CKA_VERIFY=FALSE and/or CKA_ENCRYPT=FALSE since it can not
-             * technically enforce the restrictions. Therefore override these
-             * attributes for the EP11 library, but keep the original attribute
-             * values in the object.
-             */
-            if (keytype == CKK_EC || keytype == CKK_RSA || keytype == CKK_DSA)
-                bool_value = CK_TRUE;
-            else
-                bool_value = *(CK_BBOOL *)attr->pValue;
-            rc = add_to_attribute_array(&p_attrs, &attrs_len, attr->type,
-                                        &bool_value, sizeof(bool_value));
-            if (rc != CKR_OK) {
-                TRACE_ERROR("%s adding attribute failed type=0x%lx rc=0x%lx\n",
-                            __func__, attr->type, rc);
-                goto make_maced_spki_end;
-            }
-            break;
+            node = node->next;
+        }
+    } else {
+        rc = get_ulong_attribute_by_type(pub_key_attrs, pub_key_attrs_len,
+                                         CKA_KEY_TYPE, &keytype);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("Could not find CKA_KEY_TYPE for the key.\n");
+            goto make_maced_spki_end;
+        }
 
-        case CKA_EXTRACTABLE:
-        case CKA_MODIFIABLE:
-        case CKA_DERIVE:
-        case CKA_WRAP:
-        case CKA_TRUSTED:
-        case CKA_IBM_RESTRICTABLE:
-        case CKA_IBM_NEVER_MODIFIABLE:
-        case CKA_IBM_ATTRBOUND:
-        case CKA_IBM_USE_AS_DATA:
-            if (attr->ulValueLen > 0 && attr->pValue == NULL) {
-                rc = CKR_ATTRIBUTE_VALUE_INVALID;
-                goto make_maced_spki_end;
-            }
-            rc = add_to_attribute_array(&p_attrs, &attrs_len, attr->type,
-                                        attr->pValue, attr->ulValueLen);
-            if (rc != CKR_OK) {
-                TRACE_ERROR("%s adding attribute failed type=0x%lx rc=0x%lx\n",
-                            __func__, attr->type, rc);
-                goto make_maced_spki_end;
-            }
-            break;
+        is_private = ep11_is_private_object(pub_key_attrs, pub_key_attrs_len);
+        is_session = !ep11_is_session_object(pub_key_attrs, pub_key_attrs_len);
```

Comment:
Just to double-check: is_session = !ep11_is_session_object ? Why the negation?

Reply:
hmm good finding!  I guess the negation is wrong!